### PR TITLE
Code refactor and improvements

### PR DESCRIPTION
**Pull request by n-ski**
An assortment of improvements that I can't be bothered to split into multiple smaller PRs.

**General changes**:
- Changed the signatures of `private protected` helper methods (`GetCommentSearchResult`, `GetPostSearchResult`, etc). Now instead of there being one method that takes `object` parameter, there is two overloads for each of those methods: one overload takes `JToken` parameter, the other takes `XmlNode` parameter. For classes that override these methods, that makes it clear what kind of object they want to work with. For callers, that makes it clear what kind of object they are allowed to pass.
- Fixed incorrect preview URL of `Search.Post.SearchResult` objects created by `Template.Gelbooru`.
- Moved some of the helper methods from `ABooru` class to separate utility classes in `BooruSharp.Utils` namespace.

**Networking changes**:
- Added `GetResponseAsync(string)` and `GetResponseAsync(Uri)` helper methods in `ABooru` class and pulled `GetResponseAsync(HttpRequestMessage)` helper method from `Pixiv` class up to `ABooru` class.
- Reworked `GetJsonAsync` helper method, now it requires generic type parameter and returns deserialized JSON string into the desired type, intead of the JSON string.
- Fixed `CheckAvailabilityAsync` method, now it properly throws an exception if the service is not available. The method was also made `virtual`, so subclasses can override the default behavior.
- Made sure every `HttpResponseMessage` that is created by calling `GetResponseAsync` method is properly disposed. See [this part](https://archive.is/vhodw#32%) of Steve Gordon's article about `HttpResponseMessage` for more info.
- Moved user agent string to the `TextUtils` class and improved its creation. Now user agent string returned by the utility method looks like this: *BooruSharp/x.y.z*, where *x.y.z* is the current version of BooruSharp package. Utility method also accepts a string parameter that will insert the string in user agent (for example passing `"UnitTests"` to the utility method will return *BooruSharp.UnitTests/x.y.z*).

**Pixiv changes**:
- Updated documentation: added missing exceptions and added remaks about required authorization.
- Refactored a bunch of repeated code into separate methods.
- Changed `CheckUpdateTokenAsync` method visibility to `private` since it is used like a helper method throughout the code, now it also checks that the auth tokens are set.
- Fixed `AddFavoriteAsync` and `RemoveFavoriteAsync` methods not updating auth tokens.
- Added a semaphore to make sure only one login request is being processed in case multiple login/auth requests are made. Semaphore also ensures calls to `AccessToken` and `RefreshToken` properties are blocked until current login request is fully processed and tokens are set/refreshed.

### CODE REVIEW

**Comment by KoB-Kirito** [comment]
~~You have to use the second ctor if you want the semaphore to function as a lock (new SemaphoreSlim(1,1)).~~
Me dumb.

**Comment by KoB-Kirito** [comment]
Also you don't have to dispose a slim semaphore as it has no unmanaged code unless you access the AvailableWaitHandle, and you don't in this case.
Also having pixiv as a disposable could tempt users to use it in a using block, which is really not how it should be used in most use-cases.

**Comment by n-ski** [comment]
> You have to use the second ctor if you want the semaphore to function as a lock (new SemaphoreSlim(1,1)).

That constructor is only ever useful if you want to have some of the "locks" on the semaphore initially taken. `new SemaphoreSlim(1, 1)` and `new SemaphoreSlim(1)` will have the same effect and only one thread will be able to enter semaphore. See [example code](https://dotnetfiddle.net/DJrNGz) here.

> you don't have to dispose a slim semaphore as it has no unmanaged code unless you access the AvailableWaitHandle

Will be addressed, thank you.

**Comment by KoB-Kirito** [comment]
firefox_2020-09-16_16-52-41
?

**Comment by n-ski** [comment]
Yeah, the example code is kinda flawed because calls to WriteLine aren't protected. I fixed it.

File: BooruSharp.Others/Pixiv.cs [hunk 1]
```diff
@@ -18,7 +20,7 @@ namespace BooruSharp.Others
     /// Pixiv.
     /// <para>https://www.pixiv.net/</para>
     /// </summary>
-    public class Pixiv : ABooru
+    public class Pixiv : ABooru, IDisposable
```
Comment (KoB-Kirito) [inline BooruSharp.Others/Pixiv.cs]: As stated, no IDisposable pls.

**Review by KoB-Kirito — CHANGES_REQUESTED** [review]
Disposable.

**Comment by n-ski** [comment]
Looking back at this, there's too much stuff going on (breaking a lot of public APIs and whatnot). Will probably split those changes into multiple PRs if I can think of a better way to handle them.

**Pull request closed without merging**